### ISSUE TITLE
How to work with urls that have arguments?

### ISSUE THREAD

**Issue by copyfactory** [wrabit/django-cotton#18]
I'm having a great time migrating a lot of boilerplate + dropdowns so thank you for this library!

One pattern I have not been able to crack properly is working with URLs that have arguments. 

This is what I have tried so far but for some reason it still tries rendering the URL and a NoReverseMatch error is throw when a URL is not passed. Further, I can't seem to get database IDs or other URL arguments to be passed down.

_button.cotton.html_
```
<c-vars method url base_class class=""/>

<button {{ attrs }}

  {% if url %}
  hx-{{ method }}="{% url url %}"
  {% endif %}
  
    class="inline-flex items-center gap-x-2 rounded {{ padding }} {{ base_class }} {{ class }}">
  {{ icon_left }}
  {{ slot }}
  {{ icon_right }}
</button>
```

_Usage_

```
<c-buttons.button 
   class="w-full"
    method="delete"
     <--- have tried many attempts of passing the url without success --->
    url="delete_obj {{ obj.id }}" 
    :url="delete_obj obj.id"
    :url="'delete_obj' {{ obj.id }}"
    <--- end attempts --->
    hx-confirm="Please confirm you wish to delete this object."
    type="button">
    <c-slot name="icon_left">
      <c-icons.trash class="w-4 h-4"></c-icons.trash>
    </c-slot>
    Delete
</c-buttons.button>


```

Thanks,

**Comment by wrabit** [comment]
For now, it's best as a named slot:

```html
<c-buttons.button type="button" ...>
    <c-slot name="url">{% url 'delete_obj' obj.id %}</c-slot>
    Delete
</c-buttons.button>
```

I have some ideas floating around that would allow something like `attr="{% url arg1 %}"` which I will take a look at some point soon. Thanks for feedback!

**Comment by copyfactory** [comment]
Sounds good thanks!

**Comment by wrabit** [comment]
Hi @copyfactory just an FYI I just released a new version that includes this ability now, for example:

```html
<c-buttons.button 
    type="button"
    url="{% url 'delete_obj' obj.id %}">
    Delete
</c-buttons.button>
```

https://github.com/wrabit/django-cotton/releases/tag/v.0.9.9 :)

**Comment by copyfactory** [comment]
Thank you!

**Comment by wrabit** [comment]
@copyfactory Just reaching out to ask how it's going so far using cotton especially along with htmx. Does it improve the workflow of building with django + htmx? Any drawbacks? Any specific benefits? I'm thinking about writing an article highlighting benefits on using the 3. Thanks for anything in advance!

**Comment by copyfactory** [comment]
Hey! 

Funny you ask - I just deployed 4 days ago a complete rebuild of all UI components for one of my products.

TLDR is as follows:
1. Stack is Django, Alpine, HTMX, Tailwind
2. Deleted thousands of lines of code that was more or less duplicated logic across views (think modals, dropdowns, class names, etc)
3. UI consistency improved dramatically and updating the theme is now a piece of cake.
4. Similar to asyncio it can have 'project creep' since you need to rename all template files to .cotton
5. {{ attrs }} makes integrating htmx trivial.
6. I have an {{ alpine }} slot in almost all components to stuff any alpine.js logic (since binding also uses the column character). I think smarter logic around attr parsing could be included since doing <c-component x-init='open = 'false''></> will crash due to the equals in the attr value. This is technically bad practice to have in your template anyway, but sometimes it's just nice. 

One idea I'm noodling on is how to have a better integration with Django forms. Before, I was using Django-widget-tweaks which handled the type of field, if it's required, etc. I can pass errors and field names but the value and some other form stuff are a bit trickier to access :).


Do you have an email/twitter where I can message you? Happy to expand further and help you build this out if that is something you are interested it.

**Comment by wrabit** [comment]
@copyfactory Thanks for that list! Interesting points and I think 4. And 6 could be addressed, namely:

- **.cotton.html suffix requirement**. In an attempt to be efficient, this was mainly so that not all view templates were unnecessarily compiled. I think we can review something here.

- **alpine expressions in attribute values**. I think we can manage this in the compiler.

- **consider alpine :bind shorthand**. We could allow the attribute to retain the : if no template variable could be resolved but the 2 different concepts between alpine and cotton could be confusing not to mention slight additional overhead when we just intended to use alpine. We could also switch the cotton dynamic symbol to something else. Since the package is a still young I think the breaking change is ok. Considering that cotton recommends alpine I think it's worthy of changing, just to what.

Yes would really like to discuss more your insights and see what improvements we could make, and discuss more possibilities  around the forms subject https://x.com/willrabbott

**Comment by wrabit** [comment]
@copyfactory Just to let you know, the newly released version (v0.9.12) will present a breaking change as I removed the requirement for `.cotton.html` extensions of components and views that use them. More details here: https://github.com/wrabit/django-cotton/releases/tag/v0.9.12

**Comment by copyfactory** [comment]
Hey @wrabit ! I sent you a follow on twitter but I don't believe you have DMs open? 

Tristan R is my profile if you can message me :).

> .cotton.html suffix requirement. In an attempt to be efficient, this was mainly so that not all view templates were unnecessarily compiled. I think we can review something here.

Makes sense, does it introduce any performance hits by removing this constraint? Also, I'm assuming that means we walk through the templates looking for <c-... to build the component now? 

> alpine expressions in attribute values. I think we can manage this in the compiler.

Great.

> consider alpine :bind shorthand. We could allow the attribute to retain the : if no template variable could be resolved but the 2 different concepts between alpine and cotton could be confusing not to mention slight additional overhead when we just intended to use alpine. We could also switch the cotton dynamic symbol to something else. Since the package is a still young I think the breaking change is ok. Considering that cotton recommends alpine I think it's worthy of changing, just to what.

Agreed, I think there should be a clear separation between context being passed from the view (querysets, context processors, etc) and any Alpine logic. One thought I have is that with the later release is there a difference between doing:

`<c-comp :user="request.user">`

OR 

`<c-comp user="{{ request.user }}">`

If not, the ':' could be removed since using mustache is already the Django convention. Unless the ':' is the only way to pass the variable reference down to children components?

Another idea for Alpine could be to have 'injected' slots that are always available and that do not need to be defined in the component template (just like `{{ attrs  }}` ).

This could also potentially open the door for plugins later down (HTMX/hyperscript comes to mind) to reduce boilerplate when defining your components. This could also introduce some 'magic' that is less obvious and annoying though...

**Comment by wrabit** [comment]
@copyfactory yeah it turns out DMs were off.. open now, and I couldn't find you based on "Tristan R", maybe typo?

> Makes sense, does it introduce any performance hits by removing this constraint? Also, I'm assuming that means we walk through the templates looking for <c-... to build the component now?

I think ever so slightly and it will only now be the same as just adding a new templates directory to your dirs. I think the .cotton.html. requirement meant more in much earlier versions when I was still figuring out the implementation. But without it now, it feels we've nicely shedded some baggage!

> One thought I have is that with the later release is there a difference between doing `:user="request.user"` and `user="{{ request.user }}`

There is a difference because the first way (:user) you are passing the variable by reference and then second way you are passing the value of the variable.

One thing I have considered is that we make it more django-esque when we pass a variable by dropping the quotes, so `user=request.user`.  I'm not such a longtime Django user so I can't grasp the worthiness of that proposition, maybe you have an opinion about this option?

> Another idea for Alpine could be to have 'injected' slots that are always available and that do not need to be defined in the component template (just like {{ attrs }} ).

Would this be just to resolve the issue where we can't currently declare x-init / x-data etc. on the component tag? If so, I think I can resolve that and we wouldn't need to do much more.

But I like the idea about plugins, if there is a decent benefit to having HTMLX one for example. I am not a heavy user in it so I can't think of anything that would help with productivity at this point. Again, if you have any specific ideas to put forward here, happy to listen.

**Comment by wrabit** [comment]
@copyfactory FYI latest 2 releases cover:

Attributes with hyphens now accessible as snake_case 

```html
<c-component x-data="{}" />
```

```
{{ x_data }}
```

Multi-line in attributes now permitted:

```html
<div x-data="{
    something: true
  }">
  ...
</div>
```
